saleor/checkout/utils.py

```diff
@@ -1,27 +1,619 @@
+        cart = get_anonymous_cart_from_token(token, cart_queryset)
+        user = None
+    if cart is not None:
+        return cart
+    if user:
+        return Cart(user=user)
+    return Cart()
+
+
+def get_or_create_db_cart(cart_queryset=Cart.objects.all()):
+    """Decorate view to always receive a saved cart instance.
+
+    Changes the view signature from `func(request, ...)` to
+    `func(request, cart, ...)`.
+
+    If no matching cart is found, one will be created and a cookie will be set
+    for users who are not logged in.
+    """
+    # FIXME: behave like middleware and assign cart to request instead
+    def get_cart(view):
+        @wraps(view)
+        def func(request, *args, **kwargs):
+            cart = get_or_create_cart_from_request(request, cart_queryset)
+            response = view(request, cart, *args, **kwargs)
+            if not request.user.is_authenticated:
+                set_cart_cookie(cart, response)
+            return response
+        return func
+    return get_cart
+
+
+def get_or_empty_db_cart(cart_queryset=Cart.objects.all()):
+    """Decorate view to receive a cart if one exists.
+
+    Changes the view signature from `func(request, ...)` to
+    `func(request, cart, ...)`.
+
+    If no matching cart is found, an unsaved `Cart` instance will be used.
+    """
+    # FIXME: behave like middleware and assign cart to request instead
+    def get_cart(view):
+        @wraps(view)
+        def func(request, *args, **kwargs):
+            cart = get_cart_from_request(request, cart_queryset)
+            return view(request, cart, *args, **kwargs)
+        return func
+    return get_cart
+
+
+def get_cart_data(cart, shipping_range, currency, discounts, taxes):
+    """Return a JSON-serializable representation of the cart."""
+    cart_total = None
+    local_cart_total = None
+    shipping_required = False
+    total_with_shipping = None
+    local_total_with_shipping = None
+    if cart:
+        cart_total = cart.get_subtotal(discounts, taxes)
+        local_cart_total = to_local_currency(cart_total, currency)
+        shipping_required = cart.is_shipping_required()
+        total_with_shipping = TaxedMoneyRange(
+            start=cart_total, stop=cart_total)
+        if shipping_required and shipping_range:
+            total_with_shipping = shipping_range + cart_total
+        local_total_with_shipping = to_local_currency(
+            total_with_shipping, currency)
+
+    return {
+        'cart_total': cart_total,
+        'local_cart_total': local_cart_total,
+        'shipping_required': shipping_required,
+        'total_with_shipping': total_with_shipping,
+        'local_total_with_shipping': local_total_with_shipping}
+
+
+def find_open_cart_for_user(user):
+    """Find an open cart for the given user."""
+    carts = user.carts.all()
+    if len(carts) > 1:
+        logger.warning('%s has more than one open basket', user)
+        for cart in carts[1:]:
```

Comment:
Just checking, the first cart is the newest? Or shall we order above queryset by pk/created/last modified?

Reply:
First one is the newest - we track it by Cart's `last_change` field.